lib/modem_info/modem_info_params.c

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr.h>
+#include <string.h>
+#include <stdlib.h>
+#include <cJSON.h>
+#include <modem_info.h>
+#include <at_params.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(modem_info_params);
+
+int modem_info_params_init(struct modem_param_info *modem)
+{
+	if (modem == NULL) {
+		return -EINVAL;
+	}
+
+	modem->network.cur_band.type	= MODEM_INFO_CUR_BAND;
+	modem->network.sup_band.type	= MODEM_INFO_SUP_BAND;
+	modem->network.area_code.type	= MODEM_INFO_AREA_CODE;
+	modem->network.operator.type	= MODEM_INFO_OPERATOR;
+	modem->network.mcc.type		= MODEM_INFO_MCC;
+	modem->network.mnc.type		= MODEM_INFO_MNC;
+	modem->network.cellid_hex.type	= MODEM_INFO_CELLID;
+	modem->network.ip_address.type	= MODEM_INFO_IP_ADDRESS;
+	modem->network.mode.type	= MODEM_INFO_MODE;
+	modem->network.system_mode.type	= MODEM_INFO_SYSTEMMODE;
+
+	modem->sim.uicc.type		= MODEM_INFO_UICC;
+	modem->sim.iccid.type		= MODEM_INFO_ICCID;
+
+	modem->device.modem_fw.type	= MODEM_INFO_FW_VERSION;
+	modem->device.battery.type	= MODEM_INFO_BATTERY;
+	modem->device.board		= CONFIG_BOARD;
+	modem->device.app_version	= STRINGIFY(APP_VERSION);
+
+	return 0;
+}
+
+static int check_system_mode(struct lte_param *system_mode)
+{
+	char lte_string[] = "LTE-M";
+	char nbiot_string[] = "NB-IoT";
+
+	if (system_mode == NULL) {
+		return -EINVAL;
+	}
+
+	if (system_mode->value == 1) {
+		memcpy(system_mode->string, lte_string,
+			strlen(lte_string));
+	} else if (system_mode->value == 0) {
+		memcpy(system_mode->string, nbiot_string,
+			strlen(nbiot_string));
+	} else {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int area_code_parse(struct lte_param *area_code)
+{
+	if (area_code == NULL) {
+		return -EINVAL;
+	}
+
+	area_code->string[2] = '\0';
+	area_code->value = (double)strtol(area_code->string, NULL, 16);
+
+	return 0;
+}
+
+static int mcc_mnc_parse(struct lte_param *operator,
+			 struct lte_param *mcc,
+			 struct lte_param *mnc)
+{
+	if (operator == NULL || mcc == NULL || mnc == NULL) {
+		return -EINVAL;
+	}
+
+	memcpy(mcc->string, operator->string, 3);
+
+	if (strlen(operator->string) < 6) {
+		mnc->string[0] = '0';
+		memcpy(&mnc->string[1], &operator->string[3], 2);
+	} else {
+		memcpy(&mnc->string, &operator->string[3], 3);
+	}
+
+	mcc->value = (double)strtol(mcc->string, NULL, 10);
+	mnc->value = (double)strtol(mnc->string, NULL, 10);
+
+	return 0;
+}
+
+static int cellid_to_dec(struct lte_param *cellID, double *cellID_dec)
+{
+	*cellID_dec = (double)strtol(cellID->string, NULL, 16);
+
+	return 0;
+}
+
+static int modem_data_get(struct lte_param *param)
+{
+	enum at_param_type data_type;
+	int ret = 0;
+
+	data_type = modem_info_type_get(param->type);
```

Comment:
Do you really need a function for this?
Does this function need to be public? You should avoid using assert in public function and return an error instead.

Reply:
All right, I will return an error instead. 

We could restructure the firmware architecture to work without this modem_info_type_get() feature, but I don't see any easy way to go around this right now. The abstraction layers modem_info_json and modem_info_params have no knowledge of the parameter's data type, as they are stored in a lower layer. My intuition was therefore to make them request the data type before parsing them as either a string or a number to the cJSON library.  

I'll let it be as it is if this is not a blocker, and keep this in mind for future work.